### PR TITLE
Expose metadata field in servicebus and eventhub

### PR DESCRIPTION
**Pull request by Hazhzeng**
### Issues
https://github.com/Azure/azure-functions-python-worker/issues/693

### Rational
Due to the constant change in extension schema, it is pretty hard for us to keep updating the rich type bindings (e.g. Node worker never do that, they always pass json to their customer). We should expose the **metadata** field to our customer, with a json inside.

### Implementation
Reconstruct meta.Datum item into Python object using **python_value** and **python_type** field. Cache it into a variable **_trigger_metadata_pyobj**. Expose **metadata** field to customer for Python dictionary access.

**Note**
GRPC collection_string and GRPC collection_sint64 will be serialized into json array.
GRPC Bytes and GRPC collection_bytes will be based64 encoded.

### EventHub & IoTHub Usage
For EventHub and IoTHub event, user now have access to **metadata** field, which is a Python dictionary. The metadata is bound to an event object. That says, when cardinality=one, use `event.metadata` and when cardinality=many, use `event[0].metadata`. Example as follow:
```
{
  "PartitionContext": {
    "CancellationToken": {
      "IsCancellationRequested": false,
      "CanBeCanceled": true,
      "WaitHandle": {
        "Handle": {
          "value": 3592
        },
        "SafeWaitHandle": {
          "IsInvalid": false,
          "IsClosed": false
        }
      }
    },
    "ConsumerGroupName": "$Default",
    "EventHubPath": "a-eh",
    "PartitionId": "0",
    "Owner": "865713bf-b827-465b-ad2c-66209c585869",
    "RuntimeInformation": {
      "PartitionId": "0",
      "LastSequenceNumber": 0,
      "LastEnqueuedTimeUtc": "0001-01-01T00:00:00",
      "LastEnqueuedOffset": null,
      "RetrievalTime": "0001-01-01T00:00:00"
    }
  },
  "PartitionKeyArray": [],
  "OffsetArray": [
    "25769827512"
  ],
  "SequenceNumberArray": [
    96
  ],
  "EnqueuedTimeUtcArray": [
    "2020-06-18T04:40:04.171Z"
  ],
  "PropertiesArray": [
    {}
  ],
  "SystemPropertiesArray": [
    {
      "x-opt-scheduled-enqueue-time": "2020-06-18T04:40:03.643Z",
      "x-opt-sequence-number": 96,
      "x-opt-offset": "25769827512",
      "x-opt-enqueued-time": "2020-06-18T04:40:04.171Z",
      "SequenceNumber": 96,
      "Offset": "25769827512",
      "PartitionKey": null,
      "EnqueuedTimeUtc": "2020-06-18T04:40:04.171Z"
    }
  ],
  "sys": {
    "MethodName": "EventHubAMany",
    "UtcNow": "2020-06-18T04:40:04.3098934Z",
    "RandGuid": "501abf96-5db4-417f-a13e-8bb550324939"
  }
}
```

### ServiceBus Usage
Pretty similar, use `message.metadata` to acquire the metadata. Also added two additional properties **enqueued_time_utc** and **expires_at_utc**. Example below:
```
{
  "DeliveryCount": 1,
  "LockToken": "87931fd2-39f4-415a-9fdc-adfdcbed3148",
  "ExpiresAtUtc": "2020-07-02T05:39:12.17Z",
  "EnqueuedTimeUtc": "2020-06-18T05:39:12.17Z",
  "MessageId": "87c66eaf88e84119b66a26278a7b4149",
  "ContentType": "application/json",
  "SequenceNumber": 3,
  "Label": "Microsoft.Azure.ServiceBus",
  "UserProperties": {},
  "MessageReceiver": {
    "RegisteredPlugins": [],
    "ReceiveMode": 0,
    "PrefetchCount": 0,
    "LastPeekedSequenceNumber": 0,
    "Path": "hazeng-sb-s",
    "OperationTimeout": "00:01:00",
    "ServiceBusConnection": {
      "Endpoint": "sb://hazeng.servicebus.windows.net",
      "OperationTimeout": "00:01:00",
      "RetryPolicy": {
        "MinimalBackoff": "00:00:00",
        "MaximumBackoff": "00:00:30",
        "DeltaBackoff": "00:00:03",
        "MaxRetryCount": 5,
        "IsServerBusy": false,
        "ServerBusyExceptionMessage": null
      },
      "TransportType": 0,
      "TokenProvider": {},
      "IsClosedOrClosing": false
    },
    "IsClosedOrClosing": false,
    "OwnsConnection": true,
    "ClientId": "MessageReceiver1hazeng-sb-s",
    "RetryPolicy": {
      "MinimalBackoff": "00:00:00",
      "MaximumBackoff": "00:00:30",
      "DeltaBackoff": "00:00:03",
      "MaxRetryCount": 5,
      "IsServerBusy": false,
      "ServerBusyExceptionMessage": null
    }
  },
  "sys": {
    "MethodName": "ServiceBusSMany",
    "UtcNow": "2020-06-18T05:39:12.2860411Z",
    "RandGuid": "bb38deae-cc75-49f2-89f5-96ec6eb857db"
  }
}
```

### Todo
Update documentation to promote metadata usage.

### CODE REVIEW

File: azure/functions/servicebus.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import json
```
Comment (vrdmr) [inline azure/functions/servicebus.py:1]: Curious - don't we add any license headers for this library? We can do that in another PR, just asking.

File: azure/functions/servicebus.py [hunk 2]
```diff
@@ -12,10 +13,13 @@ class ServiceBusMessage(azf_sbus.ServiceBusMessage):
     def __init__(
             self, *,
             body: bytes,
+            trigger_metadata: typing.Mapping[str, typing.Any] = None,
             content_type: typing.Optional[str] = None,
             correlation_id: typing.Optional[str] = None,
             delivery_count: typing.Optional[int] = 0,
+            enqueued_time_utc: typing.Optional[datetime.datetime] = None,
             expiration_time: typing.Optional[datetime.datetime] = None,
+            expires_at_utc: typing.Optional[datetime.datetime] = None,
```
[the remark this answers is not in the document]
Comment (Hazhzeng) [reply azure/functions/servicebus.py:22]: This is the recent schema change provided by servicebus. They no longer use **expiration_time**. However, due to some customer is still using **extensions.csproj** for old functions extension, we decide to keep both of them.

File: azure/functions/servicebus.py (hunk 1 above)
Comment (Hazhzeng) [reply azure/functions/servicebus.py:1]: This is a good catch, I can do that in a separate PR.

File: azure/functions/servicebus.py [hunk 3]
```diff
@@ -102,6 +120,26 @@ def to(self) -> typing.Optional[str]:
     def user_properties(self) -> typing.Dict[str, object]:
         return self.__user_properties
 
+    @property
+    def metadata(self) -> str:
```
[the remark this answers is not in the document]
Comment (Hazhzeng) [reply azure/functions/servicebus.py]: Yeah, this is a good catch, because I was thinking json is more intuitive for customer to view all data at once. But now I'm changing it to Python dictionary for ease of access.
